src/dinit-iostream.cc

```diff
@@ -0,0 +1,740 @@
+#include <limits>
+#include <cstdio>
+#include <cstring>
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <dinit-iostream.h>
+#include <baseproc-sys.h>
+
+// See dinit-iostream.h for inforamtion about the usage
+
+namespace {
+
+// A simple trick for calculating how much digits can be in a type with compile-time integer
+// logarithm like function.
+template <typename T> constexpr unsigned type_max_numdigits(T num = std::numeric_limits<T>::max(),
+        unsigned pow = 0)
+{
+    return (num == 0) ? pow : type_max_numdigits(num / 10, pow + 1);
+}
+
+// We should have space for sign in signed variables.
```

Comment:
Either put this comment next to what it is relevant to (i.e. line 26/28) perhaps in a shorter form, or explain what it means. Eg

// Note that we need space for the sign in signed values, so we add one to the type_max_numdigits value for signed types

Reply:
Actually, just move these declarations into the functions where they are actually used. There's really no need to have them at this namespace level.